File: algorithmique/fr-ioi/4.6_graphes/colorier_max.cpp
#include <iostream>
#include <vector>

using namespace std;

int connectedComponents (int height, int width, vector<vector<char>> &grid);

int main() {
	int height, width;
    cin >> height >> width;
    vector<vector<char>> grid;
    for (int i=0; i<height; i++) {
        vector<char> v;
        for (int j=0; j<width; j++) {
            char c; cin >> c;
            v.push_back(c);
        }
        grid.push_back(v);
    }
    int nbConnected = connectedComponents(height, width, grid);
    cout << nbConnected << "\n";
}

void DFS (int height, int width, int x, int y, vector<vector<char>> &grid) {
    vector<pair<int, int>> toVisit; // (x, y) coordinates
    toVisit.push_back(make_pair(x, y));
    while (!toVisit.empty()) {
        int x = toVisit.back().first;
        int y = toVisit.back().second;
        toVisit.pop_back();
        grid[y][x] = '#';
        for (int X: {x-1, x+1}) {
            if (0 <= X && X < width && grid[y][X] == '.') {
                toVisit.push_back(make_pair(X, y));
            }
        }
        for (int Y:{y-1, y+1}) {
            if (0 <= Y && Y < height && grid[Y][x] == '.') {
                toVisit.push_back(make_pair(x, Y));
            }
        }
    }
}

int connectedComponents (int height, int width, vector<vector<char>> &grid) {
    int r = 0;
    for (int y=0; y < height; y++) {
        for (int x = 0; x < width; x++) {
            if (grid[y][x] == '.') {
                r++; // counting a new connected Component
                DFS(height, width, x, y, grid); // marking the past places
            }
        }
    }
    return r;
}

File: algorithmique/fr-ioi/4.6_graphes/colorier_max.py
import sys

def read():
    return sys.stdin.readline().strip()

def main():
    height, width = map(int, read().split())
    grid = []
    for i in range (height) :
        grid.append(list(read()))
    print(connectedComponents(width, height, grid))

def connectedComponents (w, h, grid):
    r = 0
    for y in range (h) :
        for x in range (w) :
            if grid[y][x] == '.' :
                grid = DFS(w, h, x, y, grid)
                r += 1
    return r

def DFS (w, h, x, y, grid):
    toVisit = [(x, y)]
    while toVisit :
        x, y = toVisit.pop()
        grid[y][x] = '#'
        for X, Y in [(x, y+1), (x, y-1), (x+1, y), (x-1, y)]:
            if 0 <= X < w and 0 <= Y < h and grid[Y][X] == '.' :
                toVisit.append((X, Y))
    return grid

main()
